Core/COObject.m

```diff
@@ -1331,6 +1435,96 @@ - (void)commonDidChangeValueForProperty: (NSString *)key
 
 }
 
+/**
+ * Fast path variant of -commonDidChangeValueForProperty: that only checks
+ * the inserted/removed values
+ */
+- (void)commonDidChangeValueForProperty: (NSString *)key
+							  atIndexes: (NSIndexSet *)indexes
+							withObjects: (NSArray *)objects
+						   mutationKind: (ETCollectionMutationKind)mutationKind
+{
+	ETPropertyDescription *propertyDesc = [_entityDescription propertyDescriptionForName: key];
+	id newValue = [self serializableValueForStorageKey: key];
+	id oldValue = [self popOldCoreObjectRelationshipValueForPropertyDescription: propertyDesc];
+	
+	if (propertyDesc == nil)
+	{
+		[NSException raise: NSInvalidArgumentException
+					format: @"Property %@ is not declared in the metamodel %@ for %@",
+		 key, _entityDescription, self];
+	}
+	
+	if (![self isCoreObjectCollection: newValue])
+	{
+		// Call the slow path.
+		
+		// We rely on COUnsafeRetainedMutableArray automatic removal of duplicates,
+		// otherwise we could use the fast path for non-COPrimitiveCollection multivalues.
+		
+		[self commonDidChangeValueForProperty: key];
+		return;
+	}
+	
+	// Fast path:
+	
+	[(id <COPrimitiveCollection>)newValue setMutable: NO];
+	
+	// We must figure out which objects were added, and which were replaced or removed
+	
+	NSArray *replacedOrRemoved;
+	NSArray *addedObjects;
+	
+	if (mutationKind == ETCollectionMutationKindInsertion)
+	{
+		addedObjects = objects;
+		replacedOrRemoved = @[];
+	}
+	else if (mutationKind == ETCollectionMutationKindRemoval)
+	{
+		addedObjects = @[];
+		replacedOrRemoved = objects;
+	}
+	else if (mutationKind == ETCollectionMutationKindReplacement)
+	{
+		// FIXME: How do we know which objects were replaced if it's an unordered collection?
```

Comment:
I'd like to get rid of the snapshot of old collection values in -willChangeValueForProperty:, if possible, because it's slow. It looks like most of the time in OutlineItem -addObject: is spent creating the snapshot and deallocating it, at least when the average size of the collection is 500 items.